extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/WorkerSketchFetcherTest.java

```diff
@@ -113,20 +117,30 @@ public void tearDown() throws Exception
   public void test_submitFetcherTask_parallelFetch_workerThrowsException_shouldCancelOtherTasks() throws Exception
   {
     // Store futures in a queue
-    final Queue<ListenableFuture<ClusterByStatisticsSnapshot>> futureQueue = new ConcurrentLinkedQueue<>();
+    final Queue<Future<?>> futureQueue = new ConcurrentLinkedQueue<>();
     final List<String> workerIds = ImmutableList.of("0", "1", "2", "3");
     final CountDownLatch latch = new CountDownLatch(workerIds.size());
 
-    target = spy(new WorkerSketchFetcher(workerClient, ClusterStatisticsMergeMode.PARALLEL, 300_000_000));
+    target = spy(
+        new WorkerSketchFetcher(
+            workerClient,
+            ClusterStatisticsMergeMode.PARALLEL,
+            300_000_000,
+            executorService
+        )
+    );
+
+    // When submitting futures from the executor, add it to the list first.
+    doAnswer(invocation -> {
+      Future<?> future = spy((Future<?>) invocation.callRealMethod());
+      futureQueue.add(future);
+      return future;
+    }).when(executorService).submit(any(Runnable.class));
 
-    // When fetching snapshots, return a mock and add future to queue
     doAnswer(invocation -> {
-      ListenableFuture<ClusterByStatisticsSnapshot> snapshotListenableFuture =
-          spy(Futures.immediateFuture(mock(ClusterByStatisticsSnapshot.class)));
-      futureQueue.add(snapshotListenableFuture);
       latch.countDown();
       latch.await();
-      return snapshotListenableFuture;
+      return Futures.immediateFuture(mock(ClusterByStatisticsSnapshot.class));
     }).when(workerClient).fetchClusterByStatisticsSnapshot(any(), any(), anyInt());
```

Comment:
1. shouldn't it be `eq("1")` in the first argument, since below we're adding a different mock for `eq("2")` ? 
2. Can we return a custom future from here which is never done and also captures the `cancel` invocation to increment the cancel counter - that could remove the need for spying executor service and creating new testing constructors for fetcher? 
3. Why are we doing `latch.await()` in every thread and using sleep in the main test thread? Didn't understand that part

Reply:
Changed the answers to be more concise and clear instead of layering mocks.

I didn't quite understand the second point, if the executorService future is cancelled, it does not call cancel on the custom future, since it is unaware of it entirely. It instead throws an InterruptedException if the task has already started (which is find and handled properly in our code), however, this our custom mock would not know that it was interrupted.

The sleep in the main thread was to resolve a rare test failure where the asserts were happening before the futures could be cancelled, causing the test to fail prematurely.